### ISSUE TITLE
[Bug] Admin session gets logged out when trying to automatically connect to a BTCPay Instance

### ISSUE THREAD

**Issue by pavlenex** [btcpayserver/grandnode#2]
When you try to do automatic pairing, the session gets logged out, and data is not parsed. 

https://github.com/btcpayserver/grandnode/assets/36959754/99dbcbab-860b-4011-9f3d-8b8ed8d43012

**Comment by ndeet** [comment]
@Nisaba for context this is the same problem like you have on your instance, there it did not work for me too. The weird thing is that for me it works with my own GrandNode instance and my own demo servers mainnet.demo.btcpay.tech and testnet.demo.btcpay.tech but NOT with our official demo server at mainnet.demo.btcpayserver.org. For Pavle it also does not work with official demo and his own BTCPay Server.

I will try use chrome debuger to log the redirect and check request headers and body and stuff but it seems like that although the session id is passed in the get request it gets lost for some reason. Will update here my findings later today.

**Comment by Nisaba** [comment]
Very strange, since once again, all works well for me in my test environment and live environment...
Difficult for me to do something since I can't reproduce the issue..
BTCPay live version : BTCPay Server v1.12.4

![image](https://github.com/btcpayserver/grandnode/assets/34550856/6f9a2673-e153-4df7-bae2-7764767c8de2)

**Comment by ndeet** [comment]
I captured the debug log, I think the relevant parts, was a bit tricky as it logs everything.

Both instances run BTCPay 1.12.5, it does NOT work with https://mainnet.demo.btcpayserver.org but it works with https://testnet.demo.btcpay.tech. There is no reason why one should fail and the other not, both should have normal POST request data but maybe you can see from those debug logs what fails. You can see in the success case it adds config options and redirects to /Admin/BTCPayServer/Configure and it not working case to /admin/login.

NOT working https://mainnet.demo.btcpayserver.org

```
2024-02-26 13:48:39.427 +00:00 [INF] Request starting HTTP/1.1 POST http://grandnode.demo.btcpay.tech/Admin/BTCPayServer/GetAutomaticApiKeyConfig?ssid=65a841b51d1dc2dea75febd1&btcpayuri=https%3A%2F%2Fmainnet.demo.btcpayserver.org%2F application/x-www-form-urlencoded 917
2024-02-26 13:48:39.427 +00:00 [DBG] AcceptLanguageHeaderRequestCultureProvider returned the following unsupported cultures '["de-DE","de"]'.
2024-02-26 13:48:39.428 +00:00 [DBG] AcceptLanguageHeaderRequestCultureProvider returned the following unsupported UI Cultures '["de-DE","de"]'.
2024-02-26 13:48:39.428 +00:00 [DBG] POST requests are not supported
2024-02-26 13:48:39.428 +00:00 [DBG] POST requests are not supported
2024-02-26 13:48:39.428 +00:00 [DBG] POST requests are not supported
2024-02-26 13:48:39.429 +00:00 [DBG] 14 candidate(s) found for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.429 +00:00 [DBG] Endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)' with route pattern 'Admin/{controller=Home}/{action=Index}/{id?}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.429 +00:00 [DBG] Endpoint '{**SeName}' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.429 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.ProductController.ProductDetails (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.429 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Category (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.429 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Brand (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Collection (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Vendor (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.NewsController.NewsItem (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.BlogController.BlogPost (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.BlogController.BlogPost (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.PageController.PageDetails (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.430 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.KnowledgebaseController.KnowledgebaseArticle (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.431 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.KnowledgebaseController.ArticlesByCategory (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.431 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CourseController.Details (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:48:39.435 +00:00 [DBG] Request matched endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:48:39.435 +00:00 [DBG] AuthenticationScheme: Authentication was not authenticated.
2024-02-26 13:48:39.449 +00:00 [DBG] AuthenticationScheme: Authentication was not authenticated.
2024-02-26 13:48:39.456 +00:00 [INF] Executing endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:48:39.458 +00:00 [INF] Route matched with {area = "Admin", action = "GetAutomaticApiKeyConfig", controller = "BTCPayServer"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] GetAutomaticApiKeyConfig() on controller Payments.BTCPayServer.Controllers.BTCPayServerController (Payments.BTCPayServer).
2024-02-26 13:48:39.459 +00:00 [DBG] Execution plan of authorization filters (in the following order): ["Grand.Web.Common.Filters.AuthorizeAdminAttribute+AuthorizeAdminFilter","Grand.Web.Common.Security.Authorization.PermissionAuthorizeAttribute","Grand.Web.Common.Filters.PasswordExpiredAttribute+PasswordFilter"]
2024-02-26 13:48:39.459 +00:00 [DBG] Execution plan of resource filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter"]
2024-02-26 13:48:39.460 +00:00 [DBG] Execution plan of action filters (in the following order): ["Microsoft.AspNetCore.Mvc.Filters.ControllerActionFilter (Order: -2147483648)","Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter (Order: -3000)","Grand.Infrastructure.Validators.FluentValidationFilter","Grand.Web.Common.Filters.CustomerActivityAttribute+CustomerActivityFilter"]
2024-02-26 13:48:39.460 +00:00 [DBG] Execution plan of exception filters (in the following order): ["None"]
2024-02-26 13:48:39.460 +00:00 [DBG] Execution plan of result filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter"]
2024-02-26 13:48:39.465 +00:00 [INF] Authorization failed for the request at filter 'Grand.Web.Common.Filters.AuthorizeAdminAttribute+AuthorizeAdminFilter'.
2024-02-26 13:48:39.466 +00:00 [DBG] Found the endpoints ["Route: admin/login/"] for address Microsoft.AspNetCore.Routing.RouteValuesAddress
2024-02-26 13:48:39.466 +00:00 [DBG] Successfully processed template admin/login/ for Route: admin/login/ resulting in /admin/login and 
2024-02-26 13:48:39.467 +00:00 [DBG] Link generation succeeded for endpoints ["Route: admin/login/"] with result /admin/login
2024-02-26 13:48:39.467 +00:00 [INF] Executing RedirectToRouteResult, redirecting to /admin/login from route AdminLogin.
2024-02-26 13:48:39.467 +00:00 [INF] Executed action Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer) in 7.1277ms
2024-02-26 13:48:39.467 +00:00 [INF] Executed endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:48:39.468 +00:00 [INF] HTTP POST /Admin/BTCPayServer/GetAutomaticApiKeyConfig responded 302 in 38.8615 ms
2024-02-26 13:48:39.468 +00:00 [DBG] Connection id "0HN1MNF7D9O5U" completed keep alive response.
2024-02-26 13:48:39.469 +00:00 [DBG] Connection id "0HN1MNF7D9O5U" received FIN.
2024-02-26 13:48:39.469 +00:00 [INF] Request finished HTTP/1.1 POST http://grandnode.demo.btcpay.tech/Admin/BTCPayServer/GetAutomaticApiKeyConfig?ssid=65a841b51d1dc2dea75febd1&btcpayuri=https%3A%2F%2Fmainnet.demo.btcpayserver.org%2F application/x-www-form-urlencoded 917 - 302 0 - 41.5949ms
```

WORKING https://testnet.demo.btcpay.tech

```
2024-02-26 13:58:53.090 +00:00 [INF] Request starting HTTP/1.1 POST http://grandnode.demo.btcpay.tech/Admin/BTCPayServer/GetAutomaticApiKeyConfig?ssid=65a841b51d1dc2dea75febd1&btcpayuri=https%3A%2F%2Ftestnet.demo.btcpay.tech%2F application/x-www-form-urlencoded 917
2024-02-26 13:58:53.094 +00:00 [DBG] AcceptLanguageHeaderRequestCultureProvider returned the following unsupported cultures '["de-DE","de"]'.
2024-02-26 13:58:53.094 +00:00 [DBG] AcceptLanguageHeaderRequestCultureProvider returned the following unsupported UI Cultures '["de-DE","de"]'.
2024-02-26 13:58:53.094 +00:00 [DBG] POST requests are not supported
2024-02-26 13:58:53.094 +00:00 [DBG] POST requests are not supported
2024-02-26 13:58:53.094 +00:00 [DBG] POST requests are not supported
2024-02-26 13:58:53.095 +00:00 [DBG] 14 candidate(s) found for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.095 +00:00 [DBG] Endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)' with route pattern 'Admin/{controller=Home}/{action=Index}/{id?}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint '{**SeName}' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.ProductController.ProductDetails (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Category (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Brand (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Collection (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.096 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CatalogController.Vendor (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.NewsController.NewsItem (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.BlogController.BlogPost (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.BlogController.BlogPost (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.PageController.PageDetails (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.KnowledgebaseController.KnowledgebaseArticle (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.097 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.KnowledgebaseController.ArticlesByCategory (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.098 +00:00 [DBG] Endpoint 'Grand.Web.Controllers.CourseController.Details (Grand.Web)' with route pattern '{**SeName}' is valid for the request path '/Admin/BTCPayServer/GetAutomaticApiKeyConfig'
2024-02-26 13:58:53.101 +00:00 [DBG] Request matched endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:58:53.103 +00:00 [DBG] AuthenticationScheme: Authentication was successfully authenticated.
2024-02-26 13:58:53.117 +00:00 [DBG] AuthenticationScheme: Authentication was successfully authenticated.
2024-02-26 13:58:53.126 +00:00 [INF] Executing endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:58:53.128 +00:00 [INF] Route matched with {area = "Admin", action = "GetAutomaticApiKeyConfig", controller = "BTCPayServer"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] GetAutomaticApiKeyConfig() on controller Payments.BTCPayServer.Controllers.BTCPayServerController (Payments.BTCPayServer).
2024-02-26 13:58:53.129 +00:00 [DBG] Execution plan of authorization filters (in the following order): ["Grand.Web.Common.Filters.AuthorizeAdminAttribute+AuthorizeAdminFilter","Grand.Web.Common.Security.Authorization.PermissionAuthorizeAttribute","Grand.Web.Common.Filters.PasswordExpiredAttribute+PasswordFilter"]
2024-02-26 13:58:53.129 +00:00 [DBG] Execution plan of resource filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter"]
2024-02-26 13:58:53.130 +00:00 [DBG] Execution plan of action filters (in the following order): ["Microsoft.AspNetCore.Mvc.Filters.ControllerActionFilter (Order: -2147483648)","Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter (Order: -3000)","Grand.Infrastructure.Validators.FluentValidationFilter","Grand.Web.Common.Filters.CustomerActivityAttribute+CustomerActivityFilter"]
2024-02-26 13:58:53.130 +00:00 [DBG] Execution plan of exception filters (in the following order): ["None"]
2024-02-26 13:58:53.130 +00:00 [DBG] Execution plan of result filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter"]
2024-02-26 13:58:53.136 +00:00 [DBG] Executing controller factory for controller Payments.BTCPayServer.Controllers.BTCPayServerController (Payments.BTCPayServer)
2024-02-26 13:58:53.139 +00:00 [DBG] Executed controller factory for controller Payments.BTCPayServer.Controllers.BTCPayServerController (Payments.BTCPayServer)
2024-02-26 13:58:53.150 +00:00 [DBG] Connection id "0HN1MNF7D9OB8", Request id "0HN1MNF7D9OB8:00000002": started reading request body.
2024-02-26 13:58:53.151 +00:00 [DBG] Connection id "0HN1MNF7D9OB8", Request id "0HN1MNF7D9OB8:00000002": done reading request body.
2024-02-26 13:58:53.170 +00:00 [INF] Start processing HTTP request GET "https://testnet.demo.btcpay.tech/api/v1/stores/FV9wdqZQ1Kd6dpFc9Ac2iNJiLWjmc6croze7W9d3iqE4/webhooks"
2024-02-26 13:58:53.173 +00:00 [INF] Sending HTTP request GET "https://testnet.demo.btcpay.tech/api/v1/stores/FV9wdqZQ1Kd6dpFc9Ac2iNJiLWjmc6croze7W9d3iqE4/webhooks"
2024-02-26 13:58:54.051 +00:00 [INF] Received HTTP response headers after 869.9898ms - 200
2024-02-26 13:58:54.053 +00:00 [INF] End processing HTTP request after 885.4337ms - 200
2024-02-26 13:58:54.088 +00:00 [INF] Start processing HTTP request POST "https://testnet.demo.btcpay.tech/api/v1/stores/FV9wdqZQ1Kd6dpFc9Ac2iNJiLWjmc6croze7W9d3iqE4/webhooks"
2024-02-26 13:58:54.088 +00:00 [INF] Sending HTTP request POST "https://testnet.demo.btcpay.tech/api/v1/stores/FV9wdqZQ1Kd6dpFc9Ac2iNJiLWjmc6croze7W9d3iqE4/webhooks"
2024-02-26 13:58:54.264 +00:00 [INF] Received HTTP response headers after 175.8323ms - 200
2024-02-26 13:58:54.266 +00:00 [INF] End processing HTTP request after 177.7561ms - 200
2024-02-26 13:58:54.300 +00:00 [DBG] The temp data cookie .AspNetCore.Mvc.CookieTempDataProvider was not found.
2024-02-26 13:58:54.301 +00:00 [DBG] Found the endpoints ["Route: Admin/{controller=Home}/{action=Index}/{id?}","Route: {controller=Home}/{action=Index}/{id?}"] for address Microsoft.AspNetCore.Routing.RouteValuesAddress
2024-02-26 13:58:54.302 +00:00 [DBG] Successfully processed template Admin/{controller=Home}/{action=Index}/{id?} for Route: Admin/{controller=Home}/{action=Index}/{id?} resulting in /Admin/BTCPayServer/Configure and 
2024-02-26 13:58:54.302 +00:00 [DBG] Link generation succeeded for endpoints ["Route: Admin/{controller=Home}/{action=Index}/{id?}","Route: {controller=Home}/{action=Index}/{id?}"] with result /Admin/BTCPayServer/Configure
2024-02-26 13:58:54.302 +00:00 [INF] Executing RedirectResult, redirecting to /Admin/BTCPayServer/Configure.
2024-02-26 13:58:54.324 +00:00 [INF] Executed action Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer) in 1194.2441ms
2024-02-26 13:58:54.325 +00:00 [INF] Executed endpoint 'Payments.BTCPayServer.Controllers.BTCPayServerController.GetAutomaticApiKeyConfig (Payments.BTCPayServer)'
2024-02-26 13:58:54.325 +00:00 [INF] HTTP POST /Admin/BTCPayServer/GetAutomaticApiKeyConfig responded 302 in 1229.8784 ms
2024-02-26 13:58:54.325 +00:00 [DBG] Connection id "0HN1MNF7D9OB8" completed keep alive response.
2024-02-26 13:58:54.326 +00:00 [INF] Request finished HTTP/1.1 POST http://grandnode.demo.btcpay.tech/Admin/BTCPayServer/GetAutomaticApiKeyConfig?ssid=65a841b51d1dc2dea75febd1&btcpayuri=https%3A%2F%2Ftestnet.demo.btcpay.tech%2F application/x-www-form-urlencoded 917 - 302 0 - 1235.4772ms
```

Also here again the result on Chrome debugger, you see one time 302 redirect to /admin/login and the working one with redirect to /Admin/BTCPayServer/Configure, see Response Headers -> Location:
![Screenshot_20240226_120552](https://github.com/btcpayserver/grandnode/assets/1136761/b0d32097-f41f-49b2-b681-353e14c40fd6)

![Screenshot_20240226_150929](https://github.com/btcpayserver/grandnode/assets/1136761/4adb5a68-48e0-4209-bf7f-280b645f5756)

**Comment by ndeet** [comment]
It seems the failing part is on the lines:
`AuthenticationScheme: Authentication was not authenticated.`
and when working:
`AuthenticationScheme: Authentication was successfully authenticated.` 

But see no info why it failed, maybe you can read more out of it.

Or add some debugging on the auth part or when looking up the ssid id or add some logging if the POST data is empty (could this cause an auth error?)

**Comment by ndeet** [comment]
Another thing, Rider warns me that the used BTCPay Server library has known security issues, the CVEs where fixed early 2023, can you update to a more recent version please?
![Screenshot_20240226_121956](https://github.com/btcpayserver/grandnode/assets/1136761/46f66289-47af-494d-8bd5-343af92fada9)

**Comment by Nisaba** [comment]
fixed (PR#6)
